src/Features/Core/Portable/PullMemberUp/MemberAndDestinationValidator.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.  
+
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.PullMemberUp
+{
+    internal static class MemberAndDestinationValidator
+    {
+        internal static bool IsDestinationValid(Solution solution, INamedTypeSymbol destination, CancellationToken cancellationToken)
+        {
+            if (destination == null)
+            {
+                throw ExceptionUtilities.Unreachable;
+            }
+
+            // Make sure destination is class or interface since it could be ErrorTypeSymbol
+            if (destination.TypeKind != TypeKind.Interface && destination.TypeKind != TypeKind.Class)
+            {
+                return false;
+            }
+
+            // Don't provide any refactoring option if the destination is not in source.
+            // If the destination is generated code, also don't provide refactoring since we can't make sure if we won't break it.
+            var isDestinationInSourceAndNotGeneratedCode =
+                destination.DeclaringSyntaxReferences.Length > 0 &&
+                destination.Locations.Any(location => location.IsInSource && !solution.GetDocument(location.SourceTree).IsGeneratedCode(cancellationToken));
```

Comment:
it's unclear what the `DeclaringSyntaxReferences` check is for.  Is the .Locations check not sufficient?  If it isn't, can you document why you need both checks?